[Jiale] iP

Singa-pirate opened the pull request:
### ChatBot
> "Never gonna give you up. Never gonna let you down." - Rick Astley [(More inspiring quotes)](https://www.youtube.com/watch?v=dQw4w9WgXcQ&ab_channel=RickAstley)

ChatBot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/nus-cs2103-AY2324S1/ip/pull/371)
2. double-click it
3. add your tasks
4. let it manage your task for you :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you ***are a*** Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Singa-pirate committed:
Level-0: rename, greeting, exit

Singa-pirate committed:
Level-1: echo and bye

Singa-pirate committed:
Level-2: add, list

Singa-pirate committed:
Level-3: mark as done

Singa-pirate committed:
Level-4: todos, deadlines, events

Singa-pirate committed:
Automated Text UI Testing

Singa-pirate committed:
Level-5: error handling

Singa-pirate committed:
Level-6: delete

Singa-pirate committed:
Add Level-7: Save

Singa-pirate committed:
Merge branch 'branch-Level-7'

Singa-pirate committed:
Add Level-8: Dates and Times

Singa-pirate committed:
Merge branch 'branch-Level-8'

Singa-pirate committed:
A-MoreOOP

Singa-pirate committed:
Merge branch 'A-MoreOOP'

Singa-pirate committed:
Add Parser class to parse commands

Singa-pirate committed:
Merge branch 'A-MoreOOP'

Singa-pirate committed:
Organise into packages

Singa-pirate committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Singa-pirate committed:
Add JUnit tests

Singa-pirate committed:
Fix issues with when data directory / file not found
- Add checking so that ./data directory will be created if absent
- File permission error message will only be printed once

Singa-pirate committed:
Correct app name in gradle

Singa-pirate committed:
Add JavaDoc

Singa-pirate committed:
Tweak code to follow coding standard

Singa-pirate committed:
Add Level-9: Find

Singa-pirate committed:
Merge branch 'branch-A-JavaDoc'

Singa-pirate committed:
Merge branch 'branch-A-CodingStandard'

Singa-pirate committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	data/chatbot.txt
#	src/main/java/chatbot/Parser.java
#	src/main/java/chatbot/TaskList.java

Singa-pirate committed:
Add CheckStyle
- Configure CheckStyle in gradle
- Fix most of the style violations detected

Singa-pirate committed:
Merge branch 'branch-A-CheckStyle'

lordidiot commented on src/main/java/chatbot/ChatBot.java:
```diff
@@ -0,0 +1,117 @@
+    private void handleCommand(String command) throws ChatBotException {
+        String[] words = command.split(" ");
+        String firstWord = words[0];
+        switch (firstWord) {
```
Consider abstracting away the logic from the switch to make it more readable.

Singa-pirate replied:
Yep, better abstraction will be implemented in the future when I have time

lordidiot commented on src/main/java/chatbot/Ui.java:
```diff
@@ -0,0 +1,63 @@
+ * Class that represents User Interface which deals with user input and output.
+ */
+public class Ui {
+    private static final String line = "\t____________________________________________________________";
```
Since this is used as a constant, would be good if the naming style reflects that: `String LINE`

lordidiot commented on src/main/java/chatbot/tasks/ToDoTask.java:
```diff
@@ -0,0 +1,19 @@
+package chatbot.tasks;
+
+public class ToDoTask extends Task {
+    private static String type = "T";
```
Same as before, would be good if the naming style reflects the constant nature: `String TYPE`

lordidiot commented on src/test/java/chatbot/ParserTest.java:
```diff
@@ -0,0 +1,148 @@
+            Parser.parseEventTaskCommand(command);
+            fail();
+        } catch (MissingFieldException e) {
+            assertTrue(e instanceof EventMissingFieldException);
```
`Assertions.assertThrows` can be a good alternative to this that would result in a more concise test.

Singa-pirate replied:
Yep it would be a better way when I add testing self-defined Exceptions in the future. Thanks for the suggestion!

lordidiot commented on src/test/java/chatbot/ParserTest.java:
```diff
@@ -0,0 +1,148 @@
+import chatbot.tasks.EventTask;
+import chatbot.tasks.Task;
+
+public class ParserTest {
```
Consider renaming your test methods to be more descriptive, the recommended format:
`featureUnderTest_testScenario_expectedBehavior`

Singa-pirate replied:
I see, thanks!

lordidiot commented on src/test/java/chatbot/TaskListTest.java:
```diff
@@ -0,0 +1,83 @@
+import chatbot.tasks.EventTask;
+import chatbot.tasks.ToDoTask;
+
+
```
nit: unnecessary linebreaks?

lordidiot reviewed (commented):
I like the overall structure of the code. Just left some comments on small issues.

Singa-pirate committed:
Add javafx, GUI skeleton and initial styling

Singa-pirate committed:
Add function to echo user input

Singa-pirate committed:
Add DialogBox class

Singa-pirate committed:
Flip ChatBot response DialogBox

Singa-pirate committed:
Finish GUI for ChatBot

- Make GUI functional by handling user input with ChatBot
- Remove text-based UI and testing

Singa-pirate committed:
Merge branch 'master' into HEAD

freshcabbage123 commented on src/main/java/chatbot/gui/Gui.java:
```diff
@@ -0,0 +1,68 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.VBox;
+
+import chatbot.ChatBot;
```
I believe that the import structure can be restructured to the correct lexicographical order for 'chatbot.ChatBot' import. It should be before 'javafx.scene.layout.VBox'.

Singa-pirate replied:
iirc checkstyle gave me warning for putting package imports before java modules, but thanks for your suggestion!

freshcabbage123 commented on src/main/java/chatbot/tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    public DeadlineTask(String name, String deadline) {
```
Would it be more aligned with standard practice to define your constructors earlier in your code i.e. before the initDeadline(String deadline) function?

Singa-pirate replied:
yep noted, thanks!

freshcabbage123 commented on src/main/java/chatbot/tasks/EventTask.java:
Would it be beneficial to add in Javadoc comments here? It may be helpful to standardise across your files for future reviews and for collaboration with others 👍

freshcabbage123 commented on src/main/java/chatbot/gui/DialogBox.java:
```diff
@@ -0,0 +1,56 @@
+import javafx.scene.text.Text;
+import javafx.scene.text.TextAlignment;
+
+import java.io.IOException;
```
Similar issue as before, I believe that the import statement for 'java.io.IOException' should be in the 'STANDARD_JAVA_PACKAGE' group.

freshcabbage123 commented on src/main/java/chatbot/exceptions/ChatBotException.java:
I like that you have crafted a variety of specific exceptions to ensure the smooth execution of your programme 👍

freshcabbage123 reviewed (commented):
The code looks very well written, just some minor nitpicks for the sake of the assignment! Good job!

ji-just-ji reviewed (commented):
Looks like a lot of progress on your IP. Missing a few javadocs, just remember to add them later. And do look through the style again.

ji-just-ji commented on src/main/java/chatbot/TaskList.java:
```diff
@@ -0,0 +1,128 @@
+ * Class that represents the list of tasks the user has created.
+ */
+public class TaskList {
+    public static final int MAX_NUMBER_OF_TASKS = 100;
```
Any reason for public?

Singa-pirate replied:
yep should be changed to private static final, thanks!

ji-just-ji commented on src/main/java/chatbot/tasks/DeadlineTask.java:
```diff
@@ -0,0 +1,54 @@
+import java.time.format.FormatStyle;
+
+
+public class DeadlineTask extends Task {
```
Missing a bit of javadoc here

ji-just-ji commented on src/main/java/chatbot/tasks/Task.java:
```diff
@@ -0,0 +1,34 @@
+package chatbot.tasks;
+
+public class Task {
```
Missing javadoc here too

ji-just-ji reviewed (commented):
Seems mostly ok, missing a bit of javadoc and some access modifier issues. Others ahve covered the rest.

Singa-pirate committed:
Fix gradle mainClass mistake

Singa-pirate committed:
Add Java assertions in the code

Assertions are used in various places to annotate key assumptions in the code.

Examples of such places include:
- parse command methods in Parser.java, to verify the command in under the correct classification
- methods to handle user command, to verify the user command is not empty
- write to file method in Storage.java, to verify the data being written is not corrupted in terms of minimum length

These assertions are made to ensure the key assumptions still hold true after future developments.

Singa-pirate committed:
Merge pull request #1 from Singa-pirate/branch-A-Assertions

Add Java assertions in the code

Singa-pirate committed:
Improve code quality of the project

Some notable changes include:
- Add abstractions in handleCommand method of ChatBot.java
- Replace magic numbers with self-explanatory constants
- Use guard clauses to make happy paths more prominent

Hopefully, these changes would make the project code more
readable to other programmers.

Singa-pirate committed:
Merge pull request #2 from Singa-pirate/branch-A-CodeQuality

Improve code quality of the project

Singa-pirate committed:
Add streams to task list operations

Instead of for loops, streams can be used to do the operations
in a pipelined manner, which makes the code more readable.

Singa-pirate committed:
Merge pull request #3 from Singa-pirate/branch-A-Streams

Add streams to task list operations

Singa-pirate committed:
Add Priority function

Priority function allows the user to modify priority level of each task.
This would make it easier for them to sort the tasks in the schedule.

Current implementation includes:
- New tasks are automatically assigned with 'L' priority
- User can set priority level of a task with "priority <index> <priority>
- The priority is one of {H, M, L}.
- String representation of tasks will include the priority level

Priority levels are represented by letters instead of numbers,
to avoid confusion with the task index in the command.

Singa-pirate committed:
Merge branch 'Branch-C-Priority'

Singa-pirate committed:
Improve GUI

GUI of ChatBot was polished for better UX.

Some changes include:
- Larger width of the overall app
- Larger and better font
- Smaller profile pictures and more padding
- Background color for dialog boxes
- Different background colors for user, chatBot and error messages

User and chatBot have also been renamed, for more personality.

Singa-pirate committed:
Update docs

- Update ReadMe with a user guide explaining features
- Provide screen shot representing the UI

Singa-pirate committed:
Improve coding style

Singa-pirate committed:
Add more JavaDocs

Singa-pirate committed:
Improve JUnit tests

Singa-pirate committed:
Improve README

- Add more usage guide for running the application
- Add feature list with links to each sub-section

Singa-pirate committed:
Ensure header comments follow coding standard

Singa-pirate committed:
Fix hyperlink mistakes in README

Singa-pirate committed:
Fix issue when running app on other OS
